caluma/visibilities.py

```diff
@@ -1,31 +1,54 @@
-from graphene.utils.str_converters import to_snake_case
+import inspect
+from functools import wraps
 
 
-class BaseVisibility(object):
-    """Basic visibility classes to be extended by any visibility implementation."""
+def filter_queryset_for(node):
+    """Decorate function to define filtering of queryset of specific node."""
 
-    def get_base_queryset(self, node, queryset, info):
-        """
-        Get base queryset.
+    def decorate(fn):
+        @wraps(fn)
+        def _decorated(self, node, queryset, info):
+            return fn(self, node, queryset, info)
 
-        Override to implement default visiblity filters for all nodes.
-        """
-        return queryset
+        _decorated._filter_queryset_for = node
+        return _decorated
 
-    def get_queryset(self, node, queryset, info):
-        """
-        Get queryset of specific node.
+    return decorate
 
-        Checks for node specific implement by its snake name.
 
-        e.g. for Node WorkItem it will call `get_queryset_for_work_item` if available.
-        """
-        queryset = self.get_base_queryset(node, queryset, info)
+class BaseVisibility(object):
+    """Basic visibility classes to be extended by any visibility implementation.
+
+    In combination with the decorator `@queryset_for` a custom visibility class
+    can define filtering on basis of nodes and its subclasses.
+
+    A custom visibility class could look like this:
+    ```
+    >>> from caluma.types import Node
+    ... from caluma.form.schema import Form
+    ...
+    ... class CustomVisibility(BaseVisibility):
+    ...     @filter_queryset_for(Node)
+    ...     def filter_queryset_for_all(self, node, queryset, info):
+    ...         return queryset.filter(created_by_user=info.context.request.user.username)
+    ...
+    ...     @filter_queryset_for(Form)
+    ...     def filter_queryset_for_form(self, node, queryset, info):
+    ...         return queryset.exclude(slug='protected-form')
+    """
+
+    def __init__(self):
+        queryset_fns = inspect.getmembers(
+            self, lambda m: hasattr(m, "_filter_queryset_for")
+        )
+        self._filter_querysets_for = {
+            fn._filter_queryset_for: fn for _, fn in queryset_fns
+        }
 
-        method_name = f"get_queryset_for_{to_snake_case(node.__name__)}"
-        get_queryset_for_node = getattr(self, method_name, None)
-        if get_queryset_for_node is not None:
-            queryset = get_queryset_for_node(node, queryset, info)
+    def get_queryset(self, node, queryset, info):
+        for cls in reversed(node.mro()):
```

Comment:
Just out of curiosity: Why are you iterating in reversed MRO here? I think filtering is "additive", so shouldn't really matter

Reply:
Removed it as it makes no difference and is just more code... :)